vm/src/circuit/riscv.rs

```diff
@@ -433,6 +451,7 @@ fn parse_J(cs: &mut R1CS, J: u32) {
     cs.set_eq("J=2", "opcode=23"); // auipc
     cs.set_eq("J=3", "opcode=111"); // jal
     cs.set_eq("J=4", "opcode=103"); // jalr
+    cs.nand("f3", "opcode=103");
```

Comment:
It's not clear to me that this check, while valid, is helpful. We should be explicitly checking that the `f3` chunk corresponds to one of the instructions. 

I feel like this would be a good `debug_assert!`, but not something that should be added to the constraints themselves.

Reply:
I don't get it. Do you want an alternative prover to convince our verifier about an execution where `jalr` contains a non-zero value in bits 12-14 (inclusive)?

Or, is this constraint implied by some other existing constraints?

Reply:
I'm saying that this check is an over-approximation. The property we want is that `funct3` takes on one of a specific set of values (the left column here), not that `funct3` has a non-zero bit:

<img width="269" alt="Screenshot 2024-07-11 at 11 15 21" src="https://github.com/nexus-xyz/nexus-zkvm/assets/20324824/c73092a2-eea9-4e66-80fc-f7a18bf73ed9">

Reply:
> cs.nand("f3", "opcode=103");

This is the check making sure that f3 takes a particular value depending on the opcode.

This check doesn't do anything when "opcode=103" is false. Only when "opcode=103" is non-zero, this check forces "f3" to be zero.

Reply:
> We should be explicitly checking that the f3 chunk corresponds to one of the instructions.

We're already doing it [here](https://github.com/nexus-xyz/nexus-zkvm/pull/217/files#diff-b8c59a4777a514c97d140db23abfeafad66e6b085f22d50b09cef4fff30db3faR273).

Reply:
> this check is an over-approximation

Can you name a mal-formed JALR instruction that passes the check?

Reply:
Ugh, okay I misread this check/ the comment on the `nand` function. That makes sense, and you can ignore all of this.

Reply:
An alternative formulation would be

    cs.mul("J=4", "opcode=103", "f3=0"); // jalr

This might be weaker or equivalent to the current form in the PR.